### PR TITLE
INTL-2 Intl Migrator

### PR DESCRIPTION
**Pull request by matthewward-wf**
## Motivation
  
A set of migrators used for wrapping user facing strings in `dart:intl`'s `Intl.message` function
## Changes
  
Added migrators to help automate this process.
## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react_codemod/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react_codemod/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/executables/intl_message_migration.dart
Comment (greglittlefield-wf) [inline lib/src/executables/intl_message_migration.dart:229]: Where does this 4 come from?

File: lib/src/intl_suggestors/utils.dart
Comment (greglittlefield-wf) [inline lib/src/intl_suggestors/utils.dart:17]: It seems like this case should be also checking that the elements are empty strings

File: lib/src/intl_suggestors/utils.dart
Comment (greglittlefield-wf) [inline lib/src/intl_suggestors/utils.dart:13]: FYI you can use `is!` for these checks
```suggestion
  if (node is! StringInterpolation) return false;
```

File: lib/src/intl_suggestors/utils.dart
Comment (greglittlefield-wf) [inline lib/src/intl_suggestors/utils.dart:25]: #nit there's a non-nullable getter for this
```suggestion
  if (double.tryParse(node.value) != null) return false;
```

File: lib/src/intl_suggestors/utils.dart
Comment (greglittlefield-wf) [inline lib/src/intl_suggestors/utils.dart:250]: FYI if you assign to a local variable, the type can be promoted to non-nullable (or subtypes) based on if-checks.
```suggestion
  final staticType = prop.rightHandSide.staticType;
  if (staticType == null) return true;
  if (staticType.isDartCoreBool) return true;
  if (staticType.isDartCoreNull) return true;
```

File: lib/src/intl_suggestors/utils.dart
Comment (greglittlefield-wf) [inline lib/src/intl_suggestors/utils.dart:265]: Seems like this could reuse the logic in `quotedCamelCase`

File: lib/src/intl_suggestors/intl_migrator.dart
Comment (greglittlefield-wf) [inline lib/src/intl_suggestors/intl_migrator.dart:44]: I think it'd be clearer here if the null case wasn't handled using `toString`
```suggestion
            ?.name ?? 'null';
```

File: lib/src/intl_suggestors/utils.dart
Comment (greglittlefield-wf) [inline lib/src/intl_suggestors/utils.dart:190]: Having doc comments on this and `toNestedName` would be helpful.

File: lib/src/intl_suggestors/intl_configs_migrator.dart
Comment (greglittlefield-wf) [inline lib/src/intl_suggestors/intl_configs_migrator.dart:9]: It looks like this suggestor will operate on any class with these variables, not just config classes.

File: lib/src/intl_suggestors/constants.dart
Comment (greglittlefield-wf) [inline lib/src/intl_suggestors/constants.dart:1]: Is there a reason only certain props are being checked, as opposed to excluding props we don't want to check? Seems like there could be some false negatives.

**Review by greglittlefield-wf — APPROVED** [review]
LGTM. Spotted a few things but those can be addressed in follow-up PRs. +1

I'd like to get review from someone else on the internationalization effort before merging, though.

File: lib/src/executables/intl_message_migration.dart
Comment (matthewward-wf) [reply lib/src/executables/intl_message_migration.dart:229]:
```Dart
import "package:intl/intl.dart";

//ignore: avoid_classes_with_only_static_members
class CLASSNAME {
```

Comment (matthewward-wf) [reply lib/src/executables/intl_message_migration.dart:229]: I was doing this via a "wasModified" flag to start, but that wasn't idempotent as a subsequent run would delete the file.

Comment (greglittlefield-wf) [reply lib/src/executables/intl_message_migration.dart:229]: This seems a bit brittle. What about storing that code as a variable and checking whether the file contains only that code?

File: lib/src/intl_suggestors/intl_configs_migrator.dart
Comment (matthewward-wf) [reply lib/src/intl_suggestors/intl_configs_migrator.dart:9]: I do not understand why.   

Shouldn't this limit its scope?
```Dart
 @override
  bool shouldSkip(FileContext context) =>
      !context.path.contains('_config.dart');
```

File: lib/src/intl_suggestors/constants.dart
Comment (greglittlefield-wf) [reply lib/src/intl_suggestors/constants.dart:1]: That makes sense. That reasoning might be worth capturing as a code comment somewhere.

Though, I could still see blocking props like that (just blocking `size`, `color`, `variant` might get you pretty far). Or maybe it's not so nice in practice 🤷, just thought I'd call it out.

File: lib/src/intl_suggestors/intl_configs_migrator.dart
Comment (greglittlefield-wf) [reply lib/src/intl_suggestors/intl_configs_migrator.dart:9]: Oh I'm sorry, you're totally right. Never mind!

File: lib/src/intl_suggestors/utils.dart
Comment (matthewward-wf) [reply lib/src/intl_suggestors/utils.dart:250]: Like really hard

File: lib/src/intl_suggestors/constants.dart
Comment (matthewward-wf) [reply lib/src/intl_suggestors/constants.dart:1]: Yeah.  I think in either case it is a guessing game.  Guess to include user facing string, or guess to exclude non user facing strings.

**Comment by matthewward-wf** [comment]
Ticket for the follow up work: https://jira.atl.workiva.net/browse/INTL-4

File: lib/src/intl_suggestors/intl_configs_migrator.dart
Comment (nicbiggs-wk) [inline lib/src/intl_suggestors/intl_configs_migrator.dart:19]: Does Dart do short-circuiting in boolean expressions? If so you could probably put these into one if statement since it doesn't look like anything is an exception to it.

File: lib/src/executables/intl_message_migration.dart
Comment (robbecker-wf) [inline lib/src/executables/intl_message_migration.dart:109]:
```suggestion
  // we want that set up before we do other non-codemod things that might log.
```

File: lib/src/executables/intl_message_migration.dart
Comment (robbecker-wf) [inline lib/src/executables/intl_message_migration.dart:170]: manually reading as lines and extracting the name could be replaced by using the yaml package and parsing using a library like pubspec, pubspec_parse, etc

File: lib/src/intl_suggestors/intl_configs_migrator.dart
Comment (matthewward-wf) [reply lib/src/intl_suggestors/intl_configs_migrator.dart:19]: I just kept it simple.  I didn't want to end up with
```Dart
if (varsToCheck.contains(node.name.name) && node.body is ExpressionFunctionBody && (body as ExpressionFunctionBody).expression is SimpleStringLiteral) {
```

File: lib/src/executables/intl_message_migration.dart
Comment (matthewward-wf) [reply lib/src/executables/intl_message_migration.dart:170]: Yeah, but is it worth pulling in additional deps for 4 lines of code?

**Review by robbecker-wf — APPROVED** [review]
+1 .. some optional comments .. could be included in the planned followup.

**Comment by matthewward-wf** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

File: lib/src/intl_suggestors/utils.dart
Comment (greglittlefield-wf) [reply lib/src/intl_suggestors/utils.dart:17]: Oh, I meant in this case that `node.elements.first` and `node.elements.last` have empty `.content`. You're right that the `toString()` representations are "'" though.



For example, for the string interpolation `'$bar'`
```
.elements[0]:
  runtimeType: InterpolationStringImpl
  "toString()":  "'"
  "InterpolationString.value": ""

.elements[1]:
  runtimeType: InterpolationExpressionImpl
  "toString()":  "$bar"

.elements[2]:
  runtimeType: InterpolationStringImpl
  "toString()":  "'"
  "InterpolationString.value": ""
```

<details>
<summary>(source code for that ^)</summary>

```dart
import 'package:analyzer/dart/analysis/utilities.dart';
import 'package:analyzer/dart/ast/ast.dart';
import 'package:analyzer/dart/ast/visitor.dart';

void main() {
  final unit = parseString(content: r'''
    foo() => '$bar';
  ''').unit;

  unit.accept(MyVisitor());
}

class MyVisitor extends RecursiveAstVisitor<void> {
  @override
  void visitStringInterpolation(StringInterpolation node) {
    for (var i = 0; i < node.elements.length; i++) {
      final element = node.elements[i];
      print('.elements[$i]:');
      print('  runtimeType: ${element.runtimeType}');
      print('  "toString()":  "$element"');
      if (element is InterpolationString) {
        print('  "InterpolationString.value": "${element.value}"');
      }
      print('');
    }
  }
}
```

</details>

So, I was thinking this check would be more precise if it were:
```dart
  if (node.elements.length == 3) {
    final first = node.elements.first;
    final last = node.elements.last;
    if (first is InterpolationString && last is InterpolationString && 
        first.value.isEmpty && last.value.isEmpty) {
      ...
    }
  }
```

**Pull request merged**
